[Event Hubs] Update samples, README, changelog and JSDocs (#4675)
Azure/azure-sdk-for-js: sdk/eventhub/event-hubs/changelog.md

### 2019-08-06 5.0.0-preview.2

#### General

> - The sender is refactored to avoid the `TooManyEventListeners` warning which would occur before if too many send requests were in flight at the same time from the same sender.

Reviewer: ```suggestion
- The sender is refactored to avoid the warning around too may listeners being attached which would occur before if too many send requests were in flight at the same time from the same sender.
```

- The receiver is refactored to allow the same underlying AMQP link to be shared between streaming and batching mode. This results in seamless transition between the three different receive methods on the `EventHubConsumer`
- All time related entites have been updated to use milli seconds as the unit of time for consistency.
- New error `InsufficientCreditError` is introduced for the scenario where [rhea](https://www.npmjs.com/package/rhea) is unable to send events due to its internal buffer being full. This is a transient error and so is treated as retryable.
- The error `OperationTimeoutError` was previously mistakenly classified as an AMQP error which is now corrected. Since this can also be a transient error, it is treated as retryable.

#### Publishing events

> - Introduced the `EventDataBatch`, allowing for publication of a batch of events with known size constraint. 

Reviewer: ```suggestion
- Added method `createBatch()` on the `EventHubProducer` to create an `EventDataBatch` that can then be used to add events until the maximum size is reached. This batch object can then be used in the `send()` method to send all the added events to Event Hubs.
```

>   This allows publishers to build batches without the potential for the `MaxSizeExceededError` error when sending events. It also allows publishers with bandwidth concerns to control the size of each batch published.

Reviewer: ```suggestion
  This allows publishers to build batches without the possibility of encountering the error around the message size exceeding the supported limit when sending events. It also allows publishers with bandwidth concerns to control the size of each batch published.
```

> - Enhanced the `EventHubProducer` to allow creation of an `EventDataBatch` and to accept them for publication.

Reviewer: We can remove this and as https://github.com/Azure/azure-sdk-for-js/pull/4675/files#r310844742 will cover the same intent

```suggestion
```

### 2019-08-06 5.0.0-preview.2

#### General
- The current `Sender` does not have a provision of "awaiting" on sending a message. We need to add handlers on the `Sender` for accepted, rejected, released, modified to ensure whether the message was successfully sent. 

>   Now, we use a new `AwaitableSender` instead of `Sender` which adds the handlers internally and provides an awaitable `send()` operation.

Reviewer 2: I would rephrase the whole `awaitable sender` work to just say we made performance improvements to the `send()` method that reduces the instances of TooManyEventListeners warning.

Reviewer: > The sender is refactored to avoid the `TooManyEventListeners` warning which would occur before if too many send requests were in flight at the same time from the same sender

Confirm the name of the warning

> - Refactor receiver that allows the same amqp link to be shared between streaming and batching receivers.

Reviewer: ```suggestion
- The receiver is refactored to allow the same underlying AMQP link to be shared between streaming and batching mode. This results in seamless transition between the three different receive methods on the `EventHubConsumer`
```

- All time related entites have been updated to use milli seconds as the unit of time for consistency.

> - New error InsufficientCreditError is introduced for the scenario where rhea is unable to send events due to its internal buffer being full. This is a transient error and so is treated as retryable.

Reviewer: ```suggestion
- New error `InsufficientCreditError` is introduced for the scenario where [rhea](https://www.npmjs.com/package/rhea) is unable to send events due to its internal buffer being full. This is a transient error and so is treated as retryable.
```

> - The error OperationTimeoutError was previously mistakenly classified as an AMQP error which is now corrected. Since this can also be a transient error, it is treated as retryable.

Reviewer: ```suggestion
- The error `OperationTimeoutError` was previously mistakenly classified as an AMQP error which is now corrected. Since this can also be a transient error, it is treated as retryable.
```

> - Update the version of rhea-promise to "^1.0.0".

Reviewer: Not needed as there is no net effect on end users

> - Update the version of ws to "^7.1.1".

Reviewer: Not needed. This is used only in our samples and is a dev dependency

#### Publishing events
- Introduced the `EventDataBatch`, allowing for publication of a batch of events with known size constraint. 

>   This is intended to ensure that publishers can build batches without the potential for an error when sending and to allow publishers with bandwidth concerns to control the size of each batch published.

Reviewer: ```suggestion
  This allows publishers to build batches without the potential for the `MaxSizeExceededError` error when sending events. It also allows publishers with bandwidth concerns to control the size of each batch published.
```

confirm the name of the error above with @chradek

Reviewer 2: Should be MessageTooLargeError: https://github.com/Azure/azure-sdk-for-js/blob/0c17ebe2751f1fd8f480f76e2bd8fa8056c2c61c/sdk/eventhub/event-hubs/test/sender.spec.ts#L439

- Enhanced the `EventHubProducer` to allow creation of an `EventDataBatch` and to accept them for publication.

#### Consuming events

> - Introduced the initial concept of a new version of the `EventProcessor`, intended as a neutral framework for processing events across all partitions for a given Event Hub and in the context of a specific Consumer Group. 

Reviewer: ```suggestion
- Introduced a new class `EventProcessor` which replaces the older concept of [Event Processor Host](https://www.npmjs.com/package/@azure/event-processor-host). 
```

>   This early preview is intended to allow consumers to test the new design using a single instance that does not persist checkpoints to any durable store.

Reviewer: ```suggestion
  This early preview is intended to allow users to test the new design using a single instance of `EventProcessor`. The ability to checkpoints to a durable store will be added in future updates
```

#### Retries and timeouts

> - The properties on the RetryConfig interface have been refactored for ease of use. The new RetryOptions in it will hold configurations like the number of retries, delay between retries, per try timeout etc.

Reviewer: ```suggestion
- The properties on the `RetryOptions` interface have been renamed for ease of use. 
- New property `timeoutInMs` to configure the time to wait before declaring an attempt to have failed with `OperationTimeoutError` error which is retryable.
```

> - Support for exponential retry has been added.

Reviewer: ```suggestion
- New properties `mode` and `maxRetryDelayInMs` on `RetryOptions` to configure the exponential retry mode that is now supported
```
